File: app/src/main/java/com/gproust/sprout/ui/home/BabyPane.kt
package com.gproust.sprout.ui.home

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.BabyChangingStation
import androidx.compose.material.icons.filled.BarChart
import androidx.compose.material.icons.filled.Bedtime
import androidx.compose.material.icons.filled.Favorite
import androidx.compose.material.icons.filled.LocalDrink
import androidx.compose.material.icons.filled.Medication
import androidx.compose.material.icons.filled.Monitor
import androidx.compose.material.icons.filled.TrendingUp
import androidx.compose.material.icons.filled.WaterDrop
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedCard
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.gproust.sprout.R
import com.gproust.sprout.data.local.BreastSide
import com.gproust.sprout.ui.common.StatCard
import com.gproust.sprout.ui.common.babyAge
import com.gproust.sprout.ui.common.formatDuration
import com.gproust.sprout.ui.common.formatRelative
import com.gproust.sprout.ui.navigation.Routes

/**
 * Everything about one baby that isn't a history list: how long since each of
 * the three things, a feed you can start from here, the log shortcuts, and
 * today's totals.
 *
 * Shared deliberately. With one baby the dashboard shows this in place, so
 * nothing is a tap further than it used to be; with two or more the dashboard
 * shows [BabyCard] summaries and this becomes the baby's own tab. Keeping it
 * one composable is what makes those two arrangements the same screen.
 */
@Composable
fun BabyPane(
    summary: BabySummary,
    tracksWellbeing: Boolean,
    now: Long,
    onFeed: (BreastSide) -> Unit,
    onNavigate: (String) -> Unit,
    modifier: Modifier = Modifier,
    header: (@Composable () -> Unit)? = null,
) {
    Column(modifier) {
        if (header != null) {
            header()
        }

        SinceChips(summary, now)

        Spacer(Modifier.height(12.dp))
        QuickFeed(summary.nextSide, onFeed)

        Spacer(Modifier.height(20.dp))
        SectionLabel(stringResource(R.string.home_log))
        Spacer(Modifier.height(8.dp))
        LogGrid(tracksWellbeing, onNavigate)

        Spacer(Modifier.height(20.dp))
        SectionLabel(stringResource(R.string.home_today))
        Spacer(Modifier.height(8.dp))
        TodayRow(summary)

        Spacer(Modifier.height(8.dp))
        OutlinedButton(
            onClick = { onNavigate(Routes.STATS) },
            modifier = Modifier.fillMaxWidth(),
        ) {
            Icon(Icons.Filled.BarChart, contentDescription = null)
            Text(
                stringResource(R.string.home_see_stats),
                modifier = Modifier.padding(start = 8.dp),
            )
        }
    }
}

/**
 * One baby's line on a household dashboard: the same answers as [BabyPane],
 * minus the log grid, plus a way in to the full thing.
 *
 * The feed button lives on the card rather than anywhere shared, because which
 * baby it belongs to has to be the card you touched — not a selection made on
 * another screen.
 */
@Composable
fun BabyCard(
    summary: BabySummary,
    now: Long,
    onOpen: () -> Unit,
    onFeed: (BreastSide) -> Unit,
    modifier: Modifier = Modifier,
) {
    val context = LocalContext.current
    Card(modifier = modifier.fillMaxWidth()) {
        Column(Modifier.clickable(onClick = onOpen).padding(16.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    summary.baby.name,
                    style = MaterialTheme.typography.titleLarge,
                    fontWeight = FontWeight.Bold,
                )
                Text(
                    babyAge(context, summary.baby.birthDate, now),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.padding(start = 8.dp).weight(1f),
                )
                Icon(
                    Icons.AutoMirrored.Filled.KeyboardArrowRight,
                    contentDescription = stringResource(R.string.cd_open_baby, summary.baby.name),
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            Spacer(Modifier.height(8.dp))
            SinceChips(summary, now)

            Spacer(Modifier.height(12.dp))
            QuickFeed(summary.nextSide, onFeed)
        }
    }
}

/**
 * "Fed 2h ago · Slept 40m ago · Nappy 1h ago" — the questions the app is opened
 * to answer, in the order they get asked.
 *
 * A chip with nothing behind it is left out rather than shown empty: no feed in
 * the last week is better said by silence than by a dash.
 */
@Composable
fun SinceChips(summary: BabySummary, now: Long, modifier: Modifier = Modifier) {
    val context = LocalContext.current
    Row(modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(6.dp)) {
        summary.lastFeed?.let {
            SinceChip(
                Icons.Filled.LocalDrink,
                stringResource(R.string.home_chip_fed, formatRelative(context, it, now)),
            )
        }
        summary.lastSleep?.let {
            SinceChip(
                Icons.Filled.Bedtime,
                stringResource(R.string.home_chip_slept, formatRelative(context, it, now)),
            )
        }
        summary.lastDiaper?.let {
            SinceChip(
                Icons.Filled.BabyChangingStation,
                stringResource(R.string.home_chip_nappy, formatRelative(context, it, now)),
            )
        }
    }
}

@Composable
private fun SinceChip(icon: ImageVector, label: String) {
    Surface(
        shape = RoundedCornerShape(50),
        color = MaterialTheme.colorScheme.surface,
        border = BorderStroke(1.dp, MaterialTheme.colorScheme.outlineVariant),
    ) {
        Row(
            Modifier.padding(horizontal = 10.dp, vertical = 5.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                icon,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.size(14.dp),
            )
            Text(
                label,
                style = MaterialTheme.typography.labelMedium,
                modifier = Modifier.padding(start = 5.dp),
            )
        }
    }
}

/**
 * Start a breastfeed without going through the feeding screen first.
 *
 * [next] is the breast the last session did *not* begin on, so the suggested
 * one is filled and the other stays available beside it — the alternation is a
 * default, never a rule. With nothing nursed in the last day there is no
 * alternation to continue, and both sides are offered evenly.
 */
@Composable
fun QuickFeed(next: BreastSide?, onFeed: (BreastSide) -> Unit, modifier: Modifier = Modifier) {
    Row(modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        when (next) {
            BreastSide.LEFT, BreastSide.RIGHT -> {
                val other =
                    if (next == BreastSide.LEFT) BreastSide.RIGHT else BreastSide.LEFT
                Button(onClick = { onFeed(next) }, modifier = Modifier.weight(1f)) {
                    Text(stringResource(startLabel(next)))
                }
                OutlinedButton(onClick = { onFeed(other) }) {
                    Text(stringResource(sideLabel(other)))
                }
            }
            else -> {
                Button(
                    onClick = { onFeed(BreastSide.LEFT) },
                    modifier = Modifier.weight(1f),
                ) {
                    Text(stringResource(R.string.feeding_start_left))
                }
                Button(
                    onClick = { onFeed(BreastSide.RIGHT) },
                    modifier = Modifier.weight(1f),
                ) {
                    Text(stringResource(R.string.feeding_start_right))
                }
            }
        }
    }
}

private fun startLabel(side: BreastSide) = when (side) {
    BreastSide.RIGHT -> R.string.feeding_start_right
    else -> R.string.feeding_start_left
}

private fun sideLabel(side: BreastSide) = when (side) {
    BreastSide.RIGHT -> R.string.side_right
    BreastSide.BOTH -> R.string.side_both
    else -> R.string.side_left
}

/**
 * The six things there are to log, as tiles rather than a stack of identical
 * buttons. Wellbeing joins them only for a parent who tracks it; treatments is
 * here rather than in the bottom bar, where there was never a seat for it.
 */
@Composable
fun LogGrid(tracksWellbeing: Boolean, onNavigate: (String) -> Unit, modifier: Modifier = Modifier) {
    val tiles = buildList {
        add(Tile(R.string.nav_feed, Icons.Filled.LocalDrink, Routes.FEEDING))
        add(Tile(R.string.screen_pumping, Icons.Filled.WaterDrop, Routes.PUMPING))
        add(Tile(R.string.nav_sleep, Icons.Filled.Bedtime, Routes.SLEEP))
        add(Tile(R.string.nav_diaper, Icons.Filled.BabyChangingStation, Routes.DIAPER))
        add(Tile(R.string.nav_growth, Icons.Filled.Monitor, Routes.GROWTH))
        add(Tile(R.string.screen_treatments, Icons.Filled.Medication, Routes.TREATMENTS))
        if (tracksWellbeing) {
            add(Tile(R.string.screen_wellbeing, Icons.Filled.Favorite, Routes.HEALTH))
        }
    }
    Column(modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(8.dp)) {
        tiles.chunked(3).forEach { row ->
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                row.forEach { tile ->
                    LogTile(tile, Modifier.weight(1f), onNavigate)
                }
                // Keeps a short last row's tiles the same width as a full one's
                // rather than letting three columns become one wide button.
                repeat(3 - row.size) { Spacer(Modifier.weight(1f)) }
            }
        }
    }
}

private data class Tile(val labelRes: Int, val icon: ImageVector, val route: String)

@Composable
private fun LogTile(tile: Tile, modifier: Modifier, onNavigate: (String) -> Unit) {
    OutlinedCard(
        modifier = modifier.clickable { onNavigate(tile.route) },
    ) {
        Column(
            Modifier.fillMaxWidth().padding(vertical = 14.dp, horizontal = 4.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Icon(
                tile.icon,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.primary,
                modifier = Modifier.size(24.dp),
            )
            Text(
                stringResource(tile.labelRes),
                style = MaterialTheme.typography.labelMedium,
                modifier = Modifier.padding(top = 6.dp),
            )
        }
    }
}

@Composable
private fun TodayRow(summary: BabySummary) {
    val context = LocalContext.current
    Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
        StatCard(
            label = stringResource(R.string.stat_feeds),
            value = summary.feedsToday.toString(),
            icon = Icons.Filled.LocalDrink,
            modifier = Modifier.weight(1f),
        )
        StatCard(
            label = stringResource(R.string.stat_sleep),
            value = formatDuration(context, summary.sleepTodayMs),
            icon = Icons.Filled.Bedtime,
            modifier = Modifier.weight(1f),
        )
        StatCard(
            label = stringResource(R.string.stat_diapers),
            value = summary.diapersToday.toString(),
            icon = Icons.Filled.BabyChangingStation,
            modifier = Modifier.weight(1f),
        )
    }
}

@Composable
internal fun SectionLabel(text: String) {
    Text(
        text,
        style = MaterialTheme.typography.titleMedium,
        fontWeight = FontWeight.SemiBold,
    )
}

/**
 * The gentle growth-spurt note, shown while a baby is in — or a few days from —
 * one of the typical periods. Informational only; every baby is different.
 */
@Composable
fun GrowthSpurtNote(spurt: GrowthSpurtUi, modifier: Modifier = Modifier) {
    Card(
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.secondaryContainer,
            contentColor = MaterialTheme.colorScheme.onSecondaryContainer,
        ),
        modifier = modifier.fillMaxWidth(),
    ) {
        Row(Modifier.padding(16.dp), verticalAlignment = Alignment.CenterVertically) {
            Icon(Icons.Filled.TrendingUp, contentDescription = null)
            Column(Modifier.padding(start = 12.dp)) {
                Text(
                    stringResource(
                        if (spurt.startsSoon) R.string.growth_spurt_soon_title
                        else R.string.growth_spurt_now_title,
                    ),
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.SemiBold,
                )
                Text(
                    stringResource(
                        if (spurt.startsSoon) R.string.growth_spurt_soon_body
                        else R.string.growth_spurt_now_body,
                        spurt.ageLabel,
                    ),
                    style = MaterialTheme.typography.bodyMedium,
                    modifier = Modifier.padding(top = 2.dp),
                )
            }
        }
    }
}

File: ios/Sprout/Sources/Screens/Home/BabyPane.swift
import SproutData
import SwiftUI

/// Everything about one baby that is not a history list: how long since each of
/// the three things, a feed you can start from here, the log shortcuts, and
/// today's totals.
///
/// **Shared deliberately.** With one baby the dashboard shows this in place, so
/// nothing is a tap further than it used to be; with two or more the dashboard
/// shows ``BabyCardView`` summaries and this becomes the baby's own tab. Keeping
/// it one view is what makes those two arrangements the same screen.
struct BabyPane<Header: View>: View {
    let summary: BabySummary
    let tracksWellbeing: Bool
    let now: Int64
    let onFeed: (BreastSide) -> Void
    let onOpen: (LogDestination) -> Void
    var onShareRecord: (() -> Void)?
    @ViewBuilder var header: () -> Header

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.regular) {
            header()

            SinceChips(summary: summary, now: now)
            QuickFeed(next: summary.nextSide, onFeed: onFeed)

            SectionLabel(Str.t("home_log"))
            LogGrid(tracksWellbeing: tracksWellbeing, onOpen: onOpen)

            SectionLabel(Str.t("home_today"))
            TodayRow(summary: summary)

            Button {
                onOpen(.stats)
            } label: {
                Label(Str.t("home_see_stats"), systemImage: "chart.bar.fill")
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, Spacing.snug)
            }
            .buttonStyle(.bordered)

            // The way out to the doctor's record lives here rather than on a top
            // bar, because this pane is the one thing both arrangements have: a
            // family with one baby never opens the baby's own tab (BDR-0009),
            // and an action offered only there is one half the users cannot
            // reach.
            if let onShareRecord {
                Button(action: onShareRecord) {
                    Label(Str.t("report_screen_title"), systemImage: "square.and.arrow.up")
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, Spacing.snug)
                }
                .buttonStyle(.bordered)
            }
        }
    }
}

/// One baby's line on a household dashboard: the same answers as ``BabyPane``,
/// minus the log grid, plus a way in to the full thing.
///
/// The feed button lives on the card rather than anywhere shared, because which
/// baby it belongs to has to be the card you touched — not a selection made on
/// another screen.
struct BabyCardView: View {
    let summary: BabySummary
    let now: Int64
    let onOpen: () -> Void
    let onFeed: (BreastSide) -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.snug) {
            Button(action: onOpen) {
                HStack(alignment: .firstTextBaseline, spacing: Spacing.tight) {
                    Text(summary.baby.name)
                        .font(.title3.weight(.bold))
                        .foregroundStyle(SproutColor.onSurface)
                    Text(SproutFormat.age(birthDate: summary.baby.birthDate, now: now).text)
                        .font(.callout)
                        .foregroundStyle(SproutColor.onSurfaceVariant)
                    Spacer()
                    Image(systemName: "chevron.right")
                        .foregroundStyle(SproutColor.onSurfaceVariant)
                }
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel(Str.t("cd_open_baby", summary.baby.name))

            SinceChips(summary: summary, now: now)
            QuickFeed(next: summary.nextSide, onFeed: onFeed)
        }
        .padding(Spacing.regular)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(SproutColor.surface, in: RoundedRectangle(cornerRadius: Radius.card))
    }
}

/// "Fed 2 h ago · Slept 40 min ago · Nappy 1 h ago" — the questions the app is
/// opened to answer, in the order they get asked.
struct SinceChips: View {
    let summary: BabySummary
    let now: Int64

    var body: some View {
        FlowLayout(spacing: Spacing.tight) {
            if let fed = summary.lastFeed {
                chip(Str.t("home_chip_fed", SproutFormat.relative(fed, now: now).text))
            }
            if let slept = summary.lastSleep {
                chip(Str.t("home_chip_slept", SproutFormat.relative(slept, now: now).text))
            }
            if let nappy = summary.lastDiaper {
                chip(Str.t("home_chip_nappy", SproutFormat.relative(nappy, now: now).text))
            }
        }
    }

    private func chip(_ text: String) -> some View {
        Text(text)
            .font(.footnote)
            .padding(.horizontal, Spacing.snug)
            .padding(.vertical, Spacing.hairline + 2)
            .background(SproutColor.background, in: Capsule())
            .foregroundStyle(SproutColor.onSurfaceVariant)
    }
}

/// Start a breastfeed without going anywhere first.
///
/// When the app can say which side comes next it leads with that one and offers
/// the other quietly beside it; when it cannot — nothing nursed within a day, or
/// a session that settles nothing — both are offered equally rather than
/// guessing.
struct QuickFeed: View {
    let next: BreastSide?
    let onFeed: (BreastSide) -> Void

    var body: some View {
        HStack(spacing: Spacing.tight) {
            switch next {
            case .LEFT, .RIGHT:
                let suggested = next!
                let other: BreastSide = suggested == .LEFT ? .RIGHT : .LEFT
                Button(startLabel(suggested)) { onFeed(suggested) }
                    .buttonStyle(.borderedProminent)
                    .frame(maxWidth: .infinity)
                Button(other.label) { onFeed(other) }
                    .buttonStyle(.bordered)
            default:
                Button(Str.t("feeding_start_left")) { onFeed(.LEFT) }
                    .buttonStyle(.borderedProminent)
                    .frame(maxWidth: .infinity)
                Button(Str.t("feeding_start_right")) { onFeed(.RIGHT) }
                    .buttonStyle(.borderedProminent)
                    .frame(maxWidth: .infinity)
            }
        }
    }

    private func startLabel(_ side: BreastSide) -> String {
        side == .RIGHT ? Str.t("feeding_start_right") : Str.t("feeding_start_left")
    }
}

/// Where a log tile goes.
enum LogDestination: String, Hashable {
    case feeding, pumping, sleep, diaper, growth, treatments, wellbeing
    case stats, checkIn, report
    /// Not logs — the two the dashboard's toolbar opens, and the sync screen
    /// behind Settings. They travel through the same push so there is one way
    /// into a screen rather than two.
    case profile, settings, sync

    /// The tile's accessibility identifier, and the only handle the screenshot
    /// run has on it. An identifier and not the visible label, because the
    /// capture walks the same grid in seven languages.
    var tileIdentifier: String { "log-tile-\(rawValue)" }
}

/// The five baby logs plus the parent's two, as equals.
///
/// All of them sit here rather than in the tab bar (BDR-0010): a Material
/// navigation bar holds five, Home takes one of those seats, and the arithmetic
/// stranded treatments in a list next to pumping with nothing to say why. There
/// is no seat to lose in a grid.
struct LogGrid: View {
    let tracksWellbeing: Bool
    let onOpen: (LogDestination) -> Void

    /// A named type rather than a tuple: `ForEach` cannot destructure a tuple
    /// element into closure parameters, and `\.2` is not a key path Swift will
    /// form.
    private struct Tile: Identifiable {
        let label: String
        let symbol: String
        let destination: LogDestination

        var id: LogDestination { destination }
    }

    private var tiles: [Tile] {
        var tiles: [Tile] = [
            Tile(label: Str.t("nav_feed"), symbol: "drop.fill", destination: .feeding),
            Tile(label: Str.t("screen_pumping"), symbol: "drop.triangle.fill", destination: .pumping),
            Tile(label: Str.t("nav_sleep"), symbol: "moon.zzz.fill", destination: .sleep),
            Tile(label: Str.t("nav_diaper"), symbol: "figure.child", destination: .diaper),
            Tile(label: Str.t("nav_growth"), symbol: "ruler", destination: .growth),
            Tile(label: Str.t("screen_treatments"), symbol: "pills.fill", destination: .treatments),
        ]
        // Dropped for a parent who has turned their own tracking off; the
        // history stays, untouched.
        if tracksWellbeing {
            tiles.append(Tile(label: Str.t("screen_wellbeing"), symbol: "heart.fill", destination: .wellbeing))
        }
        return tiles
    }

    var body: some View {
        // Three columns, fixed. A short last row keeps its tiles the same width
        // as a full one's rather than letting three columns become one wide
        // button.
        LazyVGrid(
            columns: Array(repeating: GridItem(.flexible(), spacing: Spacing.tight), count: 3),
            spacing: Spacing.tight
        ) {
            ForEach(tiles) { tile in
                Button {
                    onOpen(tile.destination)
                } label: {
                    VStack(spacing: Spacing.hairline + 2) {
                        Image(systemName: tile.symbol)
                            .font(.title3)
                            .foregroundStyle(SproutColor.primary)
                        Text(tile.label)
                            .font(.caption)
                            .foregroundStyle(SproutColor.onSurface)
                            .lineLimit(2)
                            .multilineTextAlignment(.center)
                            .minimumScaleFactor(0.8)
                    }
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, Spacing.snug + 2)
                    .background(
                        RoundedRectangle(cornerRadius: Radius.control)
                            .stroke(SproutColor.outline.opacity(0.4), lineWidth: 1)
                    )
                }
                .buttonStyle(.plain)
                .accessibilityIdentifier(tile.destination.tileIdentifier)
            }
        }
    }
}

private struct TodayRow: View {
    let summary: BabySummary

    var body: some View {
        HStack(spacing: Spacing.snug) {
            StatCard(
                label: Str.t("stat_feeds"),
                value: "\(summary.feedsToday)",
                systemImage: "drop.fill"
            )
            StatCard(
                label: Str.t("stat_sleep"),
                value: SproutFormat.duration(millis: summary.sleepTodayMs).text,
                systemImage: "moon.zzz.fill"
            )
            StatCard(
                label: Str.t("stat_diapers"),
                value: "\(summary.diapersToday)",
                systemImage: "figure.child"
            )
        }
    }
}
